common/core/nesteddisplay.cpp

```diff
@@ -67,10 +133,109 @@ bool NestedDisplay::SetPowerMode(uint32_t /*power_mode*/) {
   return true;
 }
 
-bool NestedDisplay::Present(std::vector<HwcLayer *> & /*source_layers*/,
+bool NestedDisplay::Present(std::vector<HwcLayer *> &source_layers,
                             int32_t * /*retire_fence*/,
                             bool /*handle_constraints*/) {
-  // TODO implement Present
+
+  int ret = 0;
+  size_t size = source_layers.size();
+  const uint32_t* pitches;
+  const uint32_t* offsets;
+  HWCNativeHandle sf_handle;
+  size_t buffer_number = 0;
+  size_t info_size = sizeof(vm_buffer_info);
+  size_t header_size = sizeof(vm_header);
+
+  for (size_t layer_index = 0; layer_index < size; layer_index++) {
+    HwcLayer* layer = source_layers.at(layer_index);
+    if (!layer->IsVisible())
+      continue;
+
+    HwcRect<int> display_frame = layer->GetDisplayFrame();
+    sf_handle = layer->GetNativeHandle();
+    auto search = mHyperDmaExportedBuffers.find(sf_handle);
+    if(search == mHyperDmaExportedBuffers.end()){
+      DrmBuffer* buffer = new DrmBuffer();
+      buffer->InitializeFromNativeHandle(sf_handle, buffer_handler_);
+
+      if(mHyperDmaBuf_Fd > 0 && buffer->GetPrimeFD()> 0){
+        struct ioctl_hyper_dmabuf_export_remote msg;
+        memset(&msg,0,sizeof(ioctl_hyper_dmabuf_export_remote));
+
+        /* TODO: add more flexibility here, instead of hardcoded domain 0*/
+        msg.remote_domain = 0;
+        msg.dmabuf_fd = buffer->GetPrimeFD();
+
+        ret = ioctl(mHyperDmaBuf_Fd, IOCTL_HYPER_DMABUF_EXPORT_REMOTE, &msg);
+        if (ret) {
+          ETRACE("Hyper DmaBuf: Exporting hyper_dmabuf failed with error %d\n", ret);
+          return false;
+        }
+        else {
+          ETRACE("Hyper DmaBuf: Exporting hyper_dmabuf Done! 0x%x\n", msg.hid.id);
+          uint32_t HalFormat = DrmFormatToHALFormat(buffer->GetFormat());
+
+          mHyperDmaExportedBuffers[sf_handle].width = buffer->GetWidth();
+          mHyperDmaExportedBuffers[sf_handle].height = buffer->GetHeight();
+          switch (HalFormat) {
+            case HAL_PIXEL_FORMAT_RGBX_8888:
+            case HAL_PIXEL_FORMAT_RGB_888:
+            case HAL_PIXEL_FORMAT_RGB_565:
+              mHyperDmaExportedBuffers[sf_handle].format = EGL_TEXTURE_RGB;
+              break;
+            case HAL_PIXEL_FORMAT_RGBA_8888:
+            case HAL_PIXEL_FORMAT_BGRA_8888:
+              mHyperDmaExportedBuffers[sf_handle].format = EGL_TEXTURE_RGBA;
+              break;
+            case HAL_PIXEL_FORMAT_NV12_Y_TILED_INTEL:
+              mHyperDmaExportedBuffers[sf_handle].format = 0x31D8;        //this wayland specified format, hardcode here.
+              break;
+          }
+          pitches = buffer->GetPitches();
+          offsets = buffer->GetOffsets();
+          mHyperDmaExportedBuffers[sf_handle].pitch[0] = pitches[0];
+          mHyperDmaExportedBuffers[sf_handle].pitch[1] = pitches[1];
```

Comment:
Btwn I am wondering if instead of sending all this data, can we just send a handle and host should be able to import the handle and retrieve all the information ?

Reply:
currently can't use handle to transmit metadata, so I have to send all the information